// @polkadot/util-crypto/key/DeriveJunction.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeriveJunction = void 0;

var _classPrivateFieldSet2 = _interopRequireDefault(require("@babel/runtime/helpers/classPrivateFieldSet"));

var _classPrivateFieldGet2 = _interopRequireDefault(require("@babel/runtime/helpers/classPrivateFieldGet"));

var _util = require("@polkadot/util");

var _asU8a = require("../blake2/asU8a.cjs");

function _classPrivateFieldInitSpec(obj, privateMap, value) { _checkPrivateRedeclaration(obj, privateMap); privateMap.set(obj, value); }

function _checkPrivateRedeclaration(obj, privateCollection) { if (privateCollection.has(obj)) { throw new TypeError("Cannot initialize the same private elements twice on an object"); } }

const RE_NUMBER = /^\d+$/;
const JUNCTION_ID_LEN = 32;
const BN_OPTIONS = {
  bitLength: 256,
  isLe: true
};

var _chainCode = /*#__PURE__*/new WeakMap();

var _isHard = /*#__PURE__*/new WeakMap();

class DeriveJunction {
  constructor() {
    _classPrivateFieldInitSpec(this, _chainCode, {
      writable: true,
      value: new Uint8Array(32)
    });

    _classPrivateFieldInitSpec(this, _isHard, {
      writable: true,
      value: false
    });
  }

  static from(value) {
    const result = new DeriveJunction();
    const [code, isHard] = value.startsWith('/') ? [value.substr(1), true] : [value, false];
    result.soft(RE_NUMBER.test(code) ? new _util.BN(code, 10) : code);
    return isHard ? result.harden() : result;
  }

  get chainCode() {
    return (0, _classPrivateFieldGet2.default)(this, _chainCode);
  }

  get isHard() {
    return (0, _classPrivateFieldGet2.default)(this, _isHard);
  }

  get isSoft() {
    return !(0, _classPrivateFieldGet2.default)(this, _isHard);
  }

  hard(value) {
    return this.soft(value).harden();
  }

  harden() {
    (0, _classPrivateFieldSet2.default)(this, _isHard, true);
    return this;
  }

  soft(value) {
    if ((0, _util.isNumber)(value) || (0, _util.isBn)(value) || (0, _util.isBigInt)(value)) {
      return this.soft((0, _util.bnToU8a)(value, BN_OPTIONS));
    } else if ((0, _util.isHex)(value)) {
      return this.soft((0, _util.hexToU8a)(value));
    } else if ((0, _util.isString)(value)) {
      return this.soft((0, _util.compactAddLength)((0, _util.stringToU8a)(value)));
    } else if (value.length > JUNCTION_ID_LEN) {
      return this.soft((0, _asU8a.blake2AsU8a)(value));
    }

    (0, _classPrivateFieldGet2.default)(this, _chainCode).fill(0);
    (0, _classPrivateFieldGet2.default)(this, _chainCode).set(value, 0);
    return this;
  }

  soften() {
    (0, _classPrivateFieldSet2.default)(this, _isHard, false);
    return this;
  }

}

exports.DeriveJunction = DeriveJunction;
